### PR TITLE
Access control based on Patient Facility

### PR DESCRIPTION
**Pull request by aethelwulffe**
Feature addition to allow a clinic to restrict access to all patient
data based on the confluence of the `user` table schedule_facilities and
the `patient_data` 'facility' fields.

This feature allows multiple clinics to share a database, while
restricting whole popluations of patient data from view.  This has the
effect of both controlling access as well as simplifying the data for
users that have no need to see what is going on in other physical or
organizational clinical service locations.

This feature is of extreme importance for Behavioral Health,
Secure/institutional Care facilities, and Assisted Living Facility
organizations.  It is also very useful for temporary care provider
situations and many others.

This is the third attempt at implementing this feature.  Prior PR's
included data schema changes that were controversial, but those changes
have since been included in recent pull requests from other
contributors.  Rather than have the history of those discussions, this
branch was created fresh for simplification of review.

### CODE REVIEW

File: library/globals.inc.php [hunk 1]
```diff
@@ -610,6 +611,12 @@
        '0',                             // default
       xl('Restrict non-authorized users to the Schedule Facilities set in User admin.')
     ),
+        'facility_acl' => array(
+      xl('Restrict User access by Facility'),
+       'bool',                          // data type
+       '0',                             // default
+      xl('Restrict User access to patients by assigned patient facility.')
+    ),
```
Comment (teryhill) [inline library/globals.inc.php]: When the role based menus are utilized this should be a function of a setting in the user field. Just a note.

File: interface/tags_filters/start.php [hunk 2]
```diff
@@ -86,4 +86,93 @@ function update_tags_filters()
 }
 add_action( 'update_plugin', 'update_tags_filters' );
 
-//}
\ No newline at end of file
+///////////////////////////////////////////////////////////////////////////////
+////test add of facility acl stuff.
```
Comment (tmccormi) [inline interface/tags_filters/start.php]: Should this comment say "Test ..."

Comment (aethelwulffe) [reply interface/tags_filters/start.php]: No....I will change.

**Comment by aethelwulffe** [comment]
Added START and END comments for the block of functions related to User Facility Access Control.

@teryhill this *may* be an example of how to use the "do_action" function calls.  I am very sorry that I didn't have the skill necessary to get the whole tags setup moved out to a module, and do the same with this code.  I am piggybacking on top of the tags.

**Comment by aethelwulffe** [comment]
Those affects are *currently* controlled by normal ACL access restrictions.  Yes, there is some expansion needed.

**Comment by teryhill** [comment]
Could be included in the GSoC report generator project.

**Comment by aethelwulffe** [comment]
Also, the schedule_facilities selector in admin/users sucks bigtime (more stuff like this needs to be handled in other screens anyway), user facility-specific provider info screen needs grouping per-provider or per-facility, needs multiple taxonomy code stuff, and a lot of other user-config at the admin level functions.

There is also one remaining boo-boo.  Patient Demographics for the `facility` field are not set as a required field in search/add, and that means it is possible to add a patient that does not have a facility...meaning the patient gets orphaned and no-one can see it without the wonders of direct database admin tools.

**Comment by teryhill** [comment]
Make an issue for those after this is committed. See if Mig wants to do the work or one of the other GCI folks.

**Comment by tmccormi** [comment]
Those are all the responses I expected.  Having tried this before (the whole reason the Tags stuff was created, in fact) ...

**Comment by aethelwulffe** [comment]
@teryhill Will do.  I intend to take a stab at moving this stuff to the /modules directory (as well as tags...whatever they do), and will try to make some implementation for some reports or possibly (but not probably) certain billing features.

I think my main issue there was use of globals in all the includes...or something...hard to remember all the problems.  Most of them involved identifying the bits of scattered code that actually made things work vs the "other stuff"

**Comment by tmccormi** [comment]
I'll proceed with the merge...

**Comment by aethelwulffe** [comment]
At least it has a toggle...

**Comment by teryhill** [comment]
@aethelwulffe let me know how to let the system know that there is a module that needs to be used.

**Pull request merged**